src/PSR7/Validators/BodyValidator/MultipartValidation.php

```diff
@@ -0,0 +1,187 @@
+                    $encodingContentType = str_replace('*', '.*', $encodingContentType);
+                    if (! preg_match('#' . $encodingContentType . '#', $partContentType)) {
+                        throw InvalidBody::becauseBodyDoesNotMatchSchemaMultipart(
+                            $partName,
+                            $partContentType,
+                            $addr
+                        );
+                    }
+                }
+
+                // 3.2. parts headers
+                foreach ($encoding->headers as $headerName => $headerSpec) {
+                    /** @var Header $headerSpec */
+                    $headerSchema = $headerSpec->schema;
+                    $headerValue  = $part->getHeader($headerName);
+
+                    if ($headerValue === null) {
+                        throw InvalidHeaders::becauseOfMissingRequiredHeaderMupripart($partName, $headerName, $addr);
+                    }
+
+                    $validator = new SchemaValidator($this->detectValidationStrategy($message));
+                    try {
+                        $validator->validate($headerValue, $headerSchema);
+                    } catch (SchemaMismatch $e) {
+                        throw InvalidHeaders::becauseValueDoesNotMatchSchemaMultipart($partName, $headerName, $headerValue, $addr, $e);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Prepare a Multipart message body (a set of normal parts) for validation against a schema
+     *
+     * @see https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html
+     * @see https://swagger.io/docs/specification/describing-request-body/multipart-requests/
+     *
+     * @return mixed[]
+     *
+     * @throws InvalidBody
+     * @throws TypeMismatch
+     */
+    private function parseMultipartData(OperationAddress $addr, StreamedPart $document) : array
+    {
+        $multipartData = []; // a buffer to fill up with message parts
+        foreach ($document->getParts() as $i => $part) {
+            $partContentType = $part->getHeader('Content-Type');
+
+            if (preg_match('#^application/.*json$#', $partContentType)) {
+                $partBody = json_decode($part->getBody(), true);
+                if (json_last_error() !== JSON_ERROR_NONE) {
+                    throw InvalidBody::becauseBodyIsNotValidJson(json_last_error_msg(), $addr);
+                }
+            } else {
+                $partBody = $part->getBody();
+            }
+
+            // if name is not set, it should be validated with "additionalProperties" keyword
+            $multipartData[$part->getName() ?? '____' . $i] = $partBody;
+        }
+
+        return $multipartData;
+    }
+
+    private function detectEncondingContentType(Encoding $encoding, StreamedPart $part, Schema $partSchema) : ?string
+    {
+        $contentType = $encoding->contentType;
+
+        if (! $contentType) {
+            // fallback strategy:
```

Comment:
Maybe flip `if-else` for fallback strategy? it seem not obvious having fallback strategy being read first

Reply:
Good point!